File: UI_stepin/day23-thread/day23-thread/src/main/V/YFTabBar.m
//
//  YFTabBar.m
//  day20-ui-lottery
//
//  Created by apple on 15/10/16.
//  Copyright (c) 2015年 itheima. All rights reserved.
//

#import "YFTabBar.h"


#define COUNT 5



@interface YFTabBar ()



@end

@implementation YFTabBar


-(void)initUI{
    UIButton *btn;
    for(int i=0;i<COUNT;i++){
        btn=[[YFTabButton alloc] init];
        [btn setImage:[UIImage imageNamed:[NSString stringWithFormat:@"TabBar%d",i+1]] forState:UIControlStateNormal];
        [btn setImage:[UIImage imageNamed:[NSString stringWithFormat:@"TabBar%dSel",i+1]] forState:UIControlStateSelected];
        [self addSubview:btn];
        btn.tag=i;
        [btn addTarget:self action:@selector(onBtnClicked:) forControlEvents:UIControlEventTouchDown];
        if(!i){
            [self onBtnClicked:btn];
        }
    }
}
-(void)selectAtIdx:(int)idx{
    [self onBtnClicked:self.subviews[idx]];
}

-(void)onBtnClicked:(UIButton *)sender{
    [self.subviews enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        [obj setSelected:NO];
    }];
    [sender setSelected:YES];
    if(self.onTabChange){
        self.onTabChange(sender.tag);
    }
}

-(void)layoutSubviews{

    CGFloat w=self.w/COUNT,
    h=self.h;
    
    [super layoutSubviews];
    [self.subviews enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
        UIButton *btn=obj;
        btn.frame=(CGRect){idx*w,0,w,h};
    }];
}


- (instancetype)init
{
    self = [super init];
    if (self) {
        [self initUI];
    }
    return self;
}

@end


@implementation YFTabButton

-(void)setHighlighted:(BOOL)highlighted{
    
}

@end

File: projects/anquanguanli/anquanguanli/anquanguanli/src/main/V/YFTabBar.swift

//
//  YFTabBar.swift
//  day-43-microblog
//
//  Created by apple on 15/12/5.
//  Copyright © 2015年 yf. All rights reserved.
//

import UIKit

protocol TBDele:NSObjectProtocol{
    func onBClick()
}

class YFTabBar: UITabBar {
    weak var dele:TBDele?

    var clo:(()->())?
    
    
   lazy var btn:UIButton = {
       let b = UIButton()
        b.addTarget(self, action: "onBClick:", forControlEvents: UIControlEvents.TouchUpInside)
        b.setBackgroundImage(UIImage(named: "tabbar_compose_button"), forState: UIControlState.Normal)
        b.setBackgroundImage(UIImage(named: "tabbar_compose_button_highlighted"), forState: UIControlState.Highlighted)
        b.setImage(UIImage(named: "tabbar_compose_icon_add"), forState: UIControlState.Normal)
        b.setImage(UIImage(named: "tabbar_compose_icon_add_highlighted"), forState: UIControlState.Highlighted)
        b.sizeToFit()
        return b
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(btn)
        backgroundImage=iimg("tabbar_background")
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let w=self.w * 0.2
        var idx=0
        for child in subviews{
            if child.isKindOfClass(NSClassFromString("UITabBarButton")!){
                child.frame=CGRect(x: CGFloat(idx)*w, y: 0, width: w, height: h)
                idx++
                if idx==2 {
                    idx++
                }
                
            }
        }
        
        btn.center=self.ic
    }
    
    
    @objc private func onBClick(b:UIButton ){
        clo?()
        dele?.onBClick()
    }
}
